singer_sdk/streams/core.py

```diff
@@ -978,44 +1061,45 @@ def _sync_records(  # noqa C901  # too complex
             )
+
             for record_result in self.get_records(current_context):
                 if isinstance(record_result, tuple):
                     # Tuple items should be the record and the child context
                     record, child_context = record_result
                 else:
                     record = record_result
-                child_context = copy.copy(
-                    self.get_child_context(record=record, context=child_context)
-                )
-                for key, val in (state_partition_context or {}).items():
-                    # Add state context to records if not already present
-                    if key not in record:
-                        record[key] = val
-
-                # Sync children, except when primary mapper filters out the record
-                if self.stream_maps[0].get_filter_result(record):
-                    self._sync_children(child_context)
-                self._check_max_record_limit(record_count)
+                try:
+                    self._process_record(
+                        record,
+                        child_context=child_context,
+                        partition_context=state_partition_context,
+                    )
+                except InvalidStreamSortException as ex:
+                    log_sort_error(
+                        log_fn=self.logger.error,
+                        ex=ex,
+                        record_count=record_count + 1,
+                        partition_record_count=partition_record_count + 1,
+                        current_context=current_context,
+                        state_partition_context=state_partition_context,
+                        stream_name=self.name,
+                    )
+                    raise ex
+
                 if selected:
-                    if (record_count - 1) % self.STATE_MSG_FREQUENCY == 0:
+                    if (
+                        record_count - 1
+                    ) % self.STATE_MSG_FREQUENCY == 0 and write_messages:
                         self._write_state_message()
-                    self._write_record_message(record)
-                    try:
-                        self._increment_stream_state(record, context=current_context)
-                    except InvalidStreamSortException as ex:
-                        log_sort_error(
-                            log_fn=self.logger.error,
-                            ex=ex,
-                            record_count=record_count + 1,
-                            partition_record_count=partition_record_count + 1,
-                            current_context=current_context,
-                            state_partition_context=state_partition_context,
-                            stream_name=self.name,
-                        )
-                        raise ex
-
-                record_count += 1
-                partition_record_count += 1
+                    if write_messages:
```

Comment:
tbh use of this `write_messages` feels clunky, but it was what I came up with to keep parent-child functionality while supporting batches.

Reply:
I don't mind this approach at all. The only other approach that comes to mind would be to make `_write_record_message()` a noop when batching is enabled.

Reply:
@aaronsteers cool, I'll leave it as is then. We can always refactor later.